libraries/core_libs/network/src/tarcap/packets_handlers/vote_packet_handler.cpp

```diff
@@ -14,17 +14,25 @@ VotePacketHandler::VotePacketHandler(std::shared_ptr<PeersState> peers_state,
       seen_votes_(1000000, 1000) {}
 
 void VotePacketHandler::validatePacketRlpFormat(const PacketData &packet_data) const {
-  if (constexpr size_t required_size = 1; packet_data.rlp_.itemCount() != required_size) {
+  if (constexpr size_t required_size = 2; packet_data.rlp_.itemCount() != required_size) {
     throw InvalidRlpItemsCountException(packet_data.type_str_, packet_data.rlp_.itemCount(), required_size);
   }
 }
 
 void VotePacketHandler::process(const PacketData &packet_data, const std::shared_ptr<TaraxaPeer> &peer) {
   auto vote = std::make_shared<Vote>(packet_data.rlp_[0].toBytes());
+  const bool reward_vote = packet_data.rlp_[1].toInt<bool>();
   const auto vote_hash = vote->getHash();
   LOG(log_dg_) << "Received PBFT vote " << vote_hash;
 
   const auto vote_round = vote->getRound();
+  if (reward_vote || (vote_round < pbft_mgr_->getPbftRound() && vote->getType() == cert_vote_type)) {
```

Comment:
how is the other part relevant? It has to be `reward_vote`

Reply:
+ I dont get why you are skipping checks below ? There are also limitations in terms of reward vote round, you could probably also use this check  `if (!seen_votes_.insert(vote_hash)) {` - double check it first

Reply:
for (vote_round < pbft_mgr_->getPbftRound() && vote->getType() == cert_vote_type)
For example:
1. node1 is the most fastest node, finalized B1 with cert votes {v1, v2, v3}, and then go to next PBFT round.
2. node2 is slowest node,  when node2 cert vote v4 finalized B1 with cert votes {v1, v2, v3, v4}, node1 goes to next round already.
3. But the v4 is still valid reward votes for B1

Reply:
In the above example, that shows we cannot drop all votes less than current PBFT round. That still may be reward votes.
And we cannot use seen_votes for reward votes. For example:
node1 PBFT chain: B1
other nodes PBFT chain: B1 <- B2
1. node1 is 1 period behind other nodes and is sycning PBFT blocks. 
2. When other nodes gossip proposal PBFT block B3 with reward votes, node1 receive them. And node1 will drop the reward votes because node1 has not have B2 yet
3. When node1 sync B2, peer will send B2 with cert votes (same rewards vote in B3). And node1 cannot add them because the votes have seen already.

Reply:
to continue in this discussion, I need answer to this [conversation](https://github.com/Taraxa-project/taraxa-node/pull/1670/files#r860952137) because I am not sure if I am right about the rounds

Reply:
btw why you even need the `reward_vote` flag ? You are bypassing it anyway. Cannot you just check:
```
if (vote_round < pbft_mgr_->getPbftRound() && vote->getType() == cert_vote_type) {
  // process it as reward vote

  return;
} 

// process it as normal vote
```

Reply:
No. For example:
1. node1 is slow node, finalized B1 with cert votes {v1, v2, v3} in round 10
2. node2 is fast node, finalized B1 with cert votes {v1, v2, v3, v4} in round 10.
3. In the same time, node2 go to next round 11 and node1 is still in round 10
4. node2 propose B2 with reward votes {v1, v2, v3, v4}, and send reward votes(peers unknown) out
5. when node1 receive the v4, v4 round is 10 == node1 current round. But v4 is reward vote

Reply:
5. When node1 receives v4, v4 round is 10 == node1 current round, node1 can use it as cert_vote for round 10 then and everything should be fine ?

Reply:
5. If like you said, when node1 receives v4, v4 round is 10 == node1 current round, node1 will add v4 to unverified queue here https://github.com/Taraxa-project/taraxa-node/blob/reward_cert_votes/libraries/core_libs/network/src/tarcap/packets_handlers/vote_packet_handler.cpp#L51
At the moment, if node1 has enough next votes, node1 will go to next round 11. Then node1 will cleanup all votes in rounds less than round 11. v4 will be deleted

Reply:
Fixed. Remove (vote_round < pbft_mgr_->getPbftRound() && vote->getType() == cert_vote_type) and add (!seen_votes_.insert(vote_hash))